Query level nullability review and drafting thread

twof opened the pull request:
Creating this PR so folks can talk about the RFC and suggest additions and changes. If you make any changes to the document, feel free to add yourself to the top of the doc!

[Rendered RFC here](https://github.com/twof/graphql-spec/blob/QueryLevelNullability/rfcs/QueryLevelNullability.md)

Want to contribute? Here are some things we need:
- Additional use cases. How would you use this feature?
- Additional information about alternatives. How are you working around this limitation of GraphQL
- Prior art. Have similar things been proposed in the past?
- Pros and cons of making schema fields non-nullable.

We want to get feedback from a broad range of people since the impact of this RFC is somewhat large. People we would like additional feedback from:
- Someone at an organization that is using a client other than Apollo
- People who work in frontend and backend web

twof committed:
Create QueryLevelNullability.md

twof committed:
Optional -> Nullable

twof committed:
Update QueryLevelNullability.md

twof committed:
Update QueryLevelNullability.md

More of an explanation of the semantics of the swift force unwrap operator

twof committed:
Update QueryLevelNullability.md

twof committed:
Update QueryLevelNullability.md

twof committed:
spell check

twof commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,178 @@
+
+**Proposed by:** 
+- [Liz Jakubowski](<social or github link here>) - Yelp iOS
+- [Alex Reilly](<social or github link here>) - Yelp iOS
```
Example of a suggested change

```suggestion
- [Alex Reilly](https://github.com/twof/) - Yelp iOS
```

twof committed:
Update QueryLevelNullability.md

twof commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,178 @@
+write `Int("5")!`. If you do that, an exception will be thrown if the statement would return `null`.
+
+In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an
+error to the client rather than any data.
```
What kind of error is best in this situation? Should that be at the discretion of the user rather than being baked into the spec?

lizjakubowski replied:
if it were at the discretion of the user, I think it would need to be specified somehow in the syntax. is there another way the user could specify this?

I don't mind a generic `missingRequiredField` error or something along those lines -- what error is returned for the `@nonNull` directive?

twof committed:
Update QueryLevelNullability.md

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+# RFC: Query Level Nullability 
+
+**Proposed by:** 
+- [Liz Jakubowski](<social or github link here>) - Yelp iOS
```
```suggestion
- [Liz Jakubowski](https://github.com/lizjakubowski) - Yelp iOS
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+
+This mismatch creates some dissonance for developers who are currently forced into dealing the
+problems that commonly surround nullablility in codebases that otherwise do not need to deal with
+those problems. This makes large numbers of nulls tedious and time-consuming to handle.
```
```suggestion
This mismatch creates some dissonance for developers who are currently forced to deal with
problems that commonly surround nullablility. This makes large numbers of null fields tedious and
time-consuming to work with.
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+- What elements are essential to having an adequate user experience? How many elements can fail before you 
+  should just show an error message instead?
+- When any combination of elements can fail, it is very hard to anticipate all edge cases and how the app 
+  might look and work, including transitions to other screens.
```
let's say a little more about how this ties in to the proposal! partial results are definitely related to this problem, but they don't go away if the proposal is accepted. maybe what we want to get across is that the proposed syntax changes where these decisions are made, and removes boilerplate from the client?

nitnit: we should use a word other than "app" to keep this as platform-agnostic as possible

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+  might look and work, including transitions to other screens.
+
+In Yelp's GraphQL schema, almost all object fields are nullable except for those with ID type. 
+This adheres to what seems to be the [official best practice](https://graphql.org/learn/best-practices/#nullability).
```
this is a lil redundant with the first part of the problem statement, maybe we can combine them somehow and remove the Yelp-specific aspect of it. e.g. could just add `According to the [official best practice]() ...` to the original problem statement or something like that :)

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.
+On mobile platforms where codegen is used, clients have full control over the nullability of the properties on the
+generated types. Since nullability is expressed in the query rather than the schema, it's flexible enough to accommodate
+various use-cases (e.g., where the business `name` _is_ allowed to be nullable).
```
we should nix the web / mobile distinction for the real RFC, and instead phrase it like "without codegen" and "with codegen"

I don't remember the point I was trying to make with this sentence, it probably makes sense to rework this section a bit
> On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+There are times where a field can be `null`, but a feature that queries for the field will not function
+if it is `null`. For example if you are trying to render an information page for a movie, you won't
+be able to do that if the name field of the movie is missing. In that case it would be preferable
+to fail as early as possible.
```
is this a Netflix example?

lizjakubowski commented on rfcs/QueryLevelNullability.md:
````diff
@@ -0,0 +1,179 @@
+```graphql
+query GetBusinessName($encid: String!) {
+  business(encid: $encid) {
+    name!  <-- this!
````
```suggestion
    name!
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+### Use a directive such as `@non-null` instead
+This is the alternative being used at some of the companies represented in this proposal for the time being.
+However this feels like a common enough need to call for a language feature. A single language feature
+rather than using directives as a workaround would enable more unified public tooling around GraphQL.
```
I think this should be the #1 alternative, since it's probably what folks reviewing will be quick to suggest, given that making schema fields non-nullable (the current #1) is not best practice. whatcha think?

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+- [Liz Jakubowski](<social or github link here>) - Yelp iOS
+- [Sanae Rosen](<social or github link here>) - Yelp Android
+- [Mark Larah](https://github.com/magicmark) - Yelp Web
+- [Alex Reilly](<social or github link here>) - Yelp iOS
```
```suggestion
- [Alex Reilly](<social or github link here>) - Yelp iOS
- [Wei Xue](<social or github link here>) - Yelp iOS
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+- [Alex Reilly](<social or github link here>) - Yelp iOS
+
+This RFC proposes creating a syntactical construct for client developers to 
+express "nullability" in their queries.
```
```suggestion
This RFC proposes a syntactical construct for GraphQL clients to express the **nullability** of schema fields requested in a query.
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+## Definitions
+
+Nullability: A concept that exists across many programming languages (eg [Swift](https://developer.apple.com/documentation/swift/optional), [Kotlin](https://kotlinlang.org/docs/null-safety.html#nullable-types-and-non-null-types), [SQL](https://www.w3schools.com/sql/sql_notnull.asp))
+that is used to express when users can be certain that a value can or can never be `null` 
```
nit: we should use one term to refer to `(users | developers | clients)` throughout the RFC

twof replied:
Decided to use `developers` throughout

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+
+Codegen: Short for Code Generation, tools that exist for working with GraphQL which take queries and schemas as
+input and output code in a language of choice that represents data that will result from those 
+queries. Codegen tools exist for many platforms. As an example, [here's some info about Apollo's offerings.](https://github.com/apollographql/apollo-tooling#code-generation)
```
I replaced "queries" with "documents" ([spec](https://spec.graphql.org/June2018/#sec-Language.Document)), and rephased the second half a bit -- lmk whatcha think!
```suggestion
Codegen: Short for Code Generation, tools that exist for working with GraphQL which accept a schema and a set of documents as
input, and output code in a language of choice that represents the data returned by those 
operations. GraphQL codegen tools exist for many platforms. As an example, [here's some info about Apollo's offerings](https://github.com/apollographql/apollo-tooling#code-generation).
```

lizjakubowski commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+In Yelp's GraphQL schema, almost all object fields are nullable except for those with ID type. 
+This adheres to what seems to be the [official best practice](https://graphql.org/learn/best-practices/#nullability).
+
+This poses a problem for the mobile clients that use Apollo's codegen feature. The codegen provides 
```
I think I wrote this originally in the JIRA ticket, where it made more sense to name-drop Apollo in the problem statement. for the RFC, I think we should use Apollo only to illustrate concepts -- the problem applies to any GraphQL code generator that outputs result/model types

we could emphasize that any code generator would use the schema to determine whether the properties on the generated models are optional. if the best practice is to make all schema fields nullable, it follows that the properties on the generated models will be optional
```suggestion
This poses a problem for GraphQL clients that use codegen. The codegen provides 
```
similarly, for the RFC we should make sure that the problem is posed as a cross-platform issue and audit any references I made to mobile-specific stuff in the original JIRA ticket

lizjakubowski commented on rfcs/QueryLevelNullability.md:
````diff
@@ -0,0 +1,179 @@
+  }
+}
+```
+// TODO: This was rejected when it was proposed. Looking for more info as to why
````
are you ready for some super meta feedback? 😄 let's remove this code comment and leave a GitHub comment with the TODO instead (so that it's more prominent & promotes discussion)

aprilrd commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an
+error to the client rather than any data.
+
+On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.
```
Typescript has a number of codegen tools such as Apollo Codegen and [GraphQL Codegen by The Guild](https://www.graphql-code-generator.com/).

twof replied:
Thanks! This was Yelp specific. We removed it.

aprilrd commented on rfcs/QueryLevelNullability.md:
````diff
@@ -0,0 +1,179 @@
+  let name: String?
+}
+```
+In many cases, the client should error if the business `name` is null. If codegen were out of the picture,
````
It took me several rereads to understand the described use case. Is this a correct understanding? 

While the schema can have the nullable fields for valid reasons (such as federation), the client wants to decide if it accepts a null value for the result to simplify the client-side logic.

twof replied:
Yeah that's right! I took your wording since it's more generally applicable than what we had.

aprilrd reviewed (commented):
Thank you for writing this up! I much prefer `!` over `@nonNull`.

The current proposal does not describe how the server should behave when a `!` marked field is null. Shouldn't that be part of the proposal?

lizjakubowski reviewed (commented):
great work so far compiling all of the information from various docs 👏 some nits, some more high level stuff -- I can definitely help rework this stuff!

aprilrd commented:
just a thought, shouldn't the proposal be named `Operation Nonnullability` or something similar? The operator actually adds non-nullability and it should function in any operation type, not just a query.

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <jakubow4@illinois.edu>

twof commented on rfcs/QueryLevelNullability.md:
````diff
@@ -0,0 +1,179 @@
+This would follow the precedent set by Kotlin.
+
+### Make non-nullability apply recursively
+For example, everything in this tree would be non-nullable
+```graphql
+query! {
+  business(id: 4) {
+    name
+  }
+}
+```
````
@wxue Do you remember why we didn't want this?

lizjakubowski replied:
I think we should remove this as an alternative, I don't think we understand it well enough to vouch for it

twof commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,179 @@
+## 🗳️ Alternatives considered
+
+### Make Schema Fields Non-Nullable Instead
+Discussion on [this topic can be found here](https://medium.com/@calebmer/when-to-use-graphql-non-null-fields-4059337f6fc8)
```
TODO: Flesh this out with things about versioning, breaking changes, and the impact it has on partial results

twof committed:
Responded to feedback

Reworked some of the language to make it more generally applicable

twof committed:
Use non-internal PR link

lizjakubowski committed:
Rework some definitions, problem statement, alternatives considered

lizjakubowski committed:
fixups

aprilrd committed:
Add Young Min Kim to the proposal (#3)

lizjakubowski committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <hello@alex.dev>

lizjakubowski committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <hello@alex.dev>

lizjakubowski committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <hello@alex.dev>

twof committed:
Add work items section

lizjakubowski committed:
PR feedback

lizjakubowski committed:
Merge pull request #2 from twof/lizjakubowski-first-pass

Rework some definitions, problem statement, alternatives considered

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Liz Jakubowski <liz.jakubowski@gmail.com>

twof committed:
Update QueryLevelNullability.md

twof committed:
Merge pull request #4 from twof/work-items

Add work items section

captbaritone commented:
Hey! We shipped something like this in [Relay](https://relay.dev/) last half. We haven’t yet documented it in open source, but it’s in wide use inside Facebook. I’ll share a description of what we built, and our observations below.

## API

Our approach was to include a custom directive, `@required(action: LOG | NONE | THROW)` where actions have the following behavior:


* `THROW`: Relay throws an error when this field is read. Relay supports per-component fragments, so we are able to throw when the component which requires the field actually renders. Combined with React’s [Error Boundaries](https://reactjs.org/docs/error-boundaries.html), we are able to contain that error to a subtree of the UI. For example, if a required field is missing in a single item in a list, we can just replace that individual item with an error UI, or perhaps choose to simply omit that single list item.
* `LOG`: Is intended for a case where the client code can recover from the missing data, but the recovery UI is not expected to be needed. When this is encountered, Relay “bubbles” the nullness up to the parent node (so if `user.name` is null, you would get a null `user`) and logs the fact that it encountered a null value back to the server for alerting purposes.
* `NONE`: This is intended for cases where the child field is required in order for the parent object to “make sense” but the client can render a sensible UI even without the parent object. It simply bubbles the nullness to the parent node.

In Relay these the custom `@required` directive is compiled away and replaced with additional metadata in our runtime artifacts which tell us how to read the server response.

## Observations

* **The “required-ness” of a field is not just a client concern but, ideally, a view/component level concern**
    * You may have a single query which provides data for a large complex view. In one subtree of that UI the field may be required, but in a separate subtree it may not.
    * It’s possible that you fetch a large query with a field that is required by one view, but that view never ends up rendering.
    * Even if the view/component that requires the field renders, it may be possible to contain the error state to just that sub-view.
* **A single field may be referenced in multiple places in the same fragment**
    * With inline fragments it’s possible to reference the same field in multiple places. If one instance of the field is required and the other is not, the expected behavior can be ambiguous.
    * In the Relay compiler we make it a compiler error to have multiple references to the same field with different `@requried` status.
* **Inline fragments lead to some odd edge cases**
    * What does it mean for a field to be required within an inline fragment within a type condition? If the type condition matches, *and* the field is missing, we error?
    * This may lead to a case where your field may never be `null`, but it could still be *missing*. The client codegen would need to account for this.
    * In Relay we made it a compiler error to use `@required` in an inline fragment’s selection. We have not had any pushback/confusion on this choice internally.
* **This feature is popular**
    * As you’ve called out in your RFC, handling null values in the client is a major source of frustration for engineers and after introducing the new directive we saw explosive adoption.
    * The `THROW` variant (which is most similar to your proposal) is far and away the most popular action, accounting for ~80% of instances.
    * However, while handling null values is a pain for developers, erroring results in a bad experience for users. This makes explosive adoptions dangerous. We attempted to mitigate this by minimizing the impact of errors (scoping them to a UI subtree) and ensuring thrown errors were correctly logged/alerted and attributed back to the query/fragment which marked them as required.

aprilrd commented:
Thank you so much for your attention and participation. I am excited to hear that Facebook has run into the same issue and implemented a solution that works at scale. I have a couple of questions to understand your comment better:

- You described how Relay uses the directive. Does your internal API change its behavior based on the `@required` directive? If not, could you explain why?
- `A single field may be referenced in multiple places in the same fragment`. I see the ambiguity if a fragment aliases fields without arguments. But there is no ambiguity for fields with arguments. Did I understand correctly?
- I fail to understand the issue with an inline fragment with a type condition. Given this schema and this query, I believe a codegen tool can correctly generate the returned type. Am I missing anything?
```graphql
type Cat {
  name: String
}

type Fish {
  weight: String
}

union AnimalResult = Cat | Fish

type Query {
  animal: AnimalResult
}
```
```graphql
query getAnimal {
  animal {
    ... on Cat {
      name @required(action: THROW)
    }

    ... on Fish {
      weight
    }
  }
}
```
```typescript
type GetAnimalQuery = { __typename: 'Cat', name: string } | { __typename: 'Fish', weight: string | null; } | null
```

kassens commented:
> Does your internal API change its behavior based on the @required directive? If not, could you explain why?

No. There might be a potential small win in terms of transfer size or server resources by not computing some other fields, but this hasn't stood out as a large enough potential win yet.

> issue with an inline fragment with a type condition

The type generation gets more weird when union or interface types are involved: `... on Mammal { name }`. Relay forces `name` to be an optional key in this case even if `name` is a `String!` in the schema.

josephsavona commented:
> You described how Relay uses the directive. Does your internal API change its behavior based on the @required directive? If not, could you explain why?

No, our implementation is currently purely on the client, and Relay Compiler removes the `@required` directive before persisting queries to the server. We have considered implementing support for this directive on the server but have not done so yet. In general we imagine this as an optimization: if the server can determine that one fragment cannot be fulfilled due to a missing required field, the server could avoid further processing of that specific fragment - but other fragments may need overlapping data and should continue evaluation if they are unaffected by the missing required data.

One consideration here is that servers generally flatten away fragments during execution, but preserving the semantics of our `@required` implementation requires treating fragments as boundary points. Ie, a missing required field may bubble-up a null to the nearest nullable parent field or nearest fragment boundary, whichever is nearest. 

> A single field may be referenced in multiple places in the same fragment. I see the ambiguity if a fragment aliases fields without arguments. But there is no ambiguity for fields with arguments. Did I understand correctly?

Fields can have ambiguous required-ness regardless of the presence of arguments. Consider:

```
# BAD: this case is ambiguous, though there's a plausible reason you may want to do this:
fragment Preview on Media {
  # possible intent: for general media the thumbnail is optional
  thumbnail(size: 32)

  ... on Photo {
    # but if its a photo we require a thumbnail to exist
    thumbnail(size: 32) @required(action: NONE)
  }
}
```

Note that in this case the problem is (like in the previous point) the fact that (inline) fragments are spread into parents. If the results of `... on Photo { thumbnail(...) }` appeared under a different key in the response - e.g. `media.onPhoto.thumbnail` - then the above example would not be ambiguous. 

However, the following is clearly ambiguous and should be considered invalid:

```
# BAD: this can reasonably be considered invalid code
fragment Preview on Media {
  thumbnail(size: 32)
  thumbnail(size: 32) @required(action: NONE)
}
```

> I fail to understand the issue with an inline fragment with a type condition. Given this schema and this query, I believe a codegen tool can correctly generate the returned type. Am I missing anything?

The challenge is inline fragments on abstract types - because the results of inline fragments are flattened into the outer result object, developers have to check for the presence of keys to see whether the object matched the abstract type. Consider a variation of your example:

```
type Cat implements Animal, Node {
  id: ID!
  name: String
}

# ... other types implementing Animal

interface Animal extends Node {
  species: String
}
interface Node {
  id: ID!
}

type Query {
  node: Node
}

query getAnimal {
  node {
    ... on Animal {
      species @required
    }
  }
}
type GetAnimalQuery = 
  # node exists, is an Animal, has species: everything is non-nullable
  {
    node: {
      species: string, 
    }
  }
  # or node doesn't exist
  # or node exists and is an Animal, but species was null so the null bubbled-up and nulled out the node
  | {
    node: null
  }
  # or the node exists but is not an Animal, nothing to select
  | { }
  # or node had an error, which nulled out the parent
  | null
```

Here is _is_ possible to define types that represent the possible outcomes, but the possible combinations grow quickly as users add more required fields within more different type conditions (inline fragments).

fotoetienne commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,269 @@
+
+This syntax consciously does not cover the following use cases:
+
+- **Default Values**
```
Defaults can be quite useful from a resilience standpoint. It is relevant to this proposal in that the `!` syntax would't be as amenable to specifying a default, whereas a directive (e.g. `@nonNull(default: false)`) would open us up to that.

benjie replied:
Though replacing `!` with `! = “default”` like in variable defaults could work. This would either have to be evaluated client side (like in Relay) or the same default would have to apply to all references at the same location (whether from fragments or directly).

josephsavona commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,269 @@
+you want to ensure that the statement does not return `null` you can instead write `Int("5")!`. If you do that, an
+exception will be thrown if the statement would evaluate to `null`.
+
+In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an error
+to the client rather than any data.
```
Can you expand here? My reading of this is that if the value of a field with `!` is null at runtime, then the GraphQL executor must treat that identically to the case that the field resolver raised an exception, including returning an entry in the top-level `errors` list and nulling out the parent field (which may need to then bubble up an error if that field was non-nullable in the schema, or also had `!` applied).

This would mean that in a query such as the following, `other_media_field` being null at runtime would cause the entire query to return effectively no data (just `{data: {media: null}, errors: [...]}`). This would prevent any part of the UI from rendering because a single field was missing for one part:

```
query MediaQuery {
  media(id: "...") {
     name
     thumbnail
     ...MediaFragment
  }
}
fragment MediaFragment on Media {
  other_media_field!
}
```

How are you thinking about handling cases such as this?

kassens replied:
I read this as any field in the response that would be `null` and has a `!` on the field in the operation would cause the whole operation to return an error and no data. Even if that field is in some named fragment.

I think this would limit the usefulness extremely, basically to very simple queries. To take Yelp as an example, I don't think it'd be a good idea to error a search query if a single search result item failed to generate a star rating (assuming that's required to render a search result item).

It would be good to clarify what the behavior is if the `!` is in a named fragment.

fotoetienne replied:
At Netflix, we have implemented this on the server (using a directive syntax) and defined it to have identical semantics as the existing SDL [Non-Null behavior](https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullabilit). This has worked well.

I've proposed a [clarification](https://github.com/twof/graphql-spec/pull/5) to specify this.

josephsavona replied:
@fotoetienne Thanks for adding the clarification, that confirms my interpretation and brings us back to my original question:

How are you thinking about the fact that a single fragment with a `!` field can cause data for other fragments (or in the extreme, an entire request) to be nulled out? As @captbaritone mentioned above, one of our goals / requirements when designing `@required` was that a "missing" required value for one part of a UI cannot cause data for other parts of the UI to be nulled out. The design here violates that constraint (it breaks fragment modularity).

fotoetienne replied:
@josephsavona That's a great question. Our current server logic will make a field act as Non-Null is there _any_ Non-Null designation for that field in the query. This query example:

```graphql
fragment Preview on Media {
  thumbnail(size: 32)

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

would essentially be interpreted by the server as:

```graphql
fragment Preview on Media {
  thumbnail(size: 32) @nonNull

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

It is at least non-ambiguous, but I can see how it is a challenge for fragment modularity.

josephsavona replied:
@fotoetienne Thanks for clarifying. That seems like a severe limitation that would make applications significantly less robust than they can and arguably should be. How do you manage this in practice? Do you encourage fields with `!` to be given unique aliases so that you prevent removing data for other parts of a UI?

fotoetienne replied:
Short answer - it hasn't come up _for us_ yet, but it's an important edge-case to consider.

Long answer - Fragment modularity is particularly crucial for clients that rely heavily on components tied to individually defined Fragments that are compiled into large queries. We don't currently have many teams using that approach.

I agree that this is an important edge-case that we should think through some more. We want a design that will lead UI developers to fall into "the pit of success" rather than allowing surprising behavior like you've described.

fotoetienne replied:
An initial thought, probably flawed: reversing the server-defined logic might help. Only honor nonNull if _all_ references to that field specify it. This would, however, require client codegen to be aware of an entire query and only make types nonNull if all fragments in the query specify nonNull. Thoughts?

fotoetienne replied:
Making conflicting definitions between fragments be a validation error (like you did with @requires) is the safest approach. I like that. It's probably an acceptable limitation that you just can't have differing nullability for the same field.

kassens replied:
@joesavona even unique aliases would still null out the object by bubbling up to the parent. It's really the same as any specific field being non-nullable in the server schema and ending up null which bubbles up.

fotoetienne commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,269 @@
+
+While the schema can have nullable fields for valid reasons (such as federation), in some cases the client wants to decide
+if it accepts a `null` value for the result to simplify the client-side logic. In addition, syntax for this concept
+would allow codegen tooling to generate model types that are more ergonomic to work with, since the since the model
```
```suggestion
would allow codegen tooling to generate model types that are more ergonomic to work with, since the model
```

fotoetienne commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,269 @@
+
+## 📜 Problem Statement
+
+The two modern languages used on mobile clients, Swift and Kotlin, are both non-null by default. By contrast, in a
```
I would consider reducing the focus on specific client languages.

The crux of the problem is that the SDL nonNull (!) eliminates the possibility of partial failure on a given type. This forces the SDL author to decide for which fields partial failure is acceptable. A GraphQL schema author may not be in the best position to decide whether partial failure is acceptable for a given canvas. Additionally, the answer to whether a missing value is acceptable may vary between UI canvases. Specifying nonNull (aka. required) in the _query_ gives the UI developer the power to decide where partial failure is acceptable, and the flexibility for different canvases to have different requirements in this regard.

The pains are more noticeable in languages with first-class null safety, but are relevant to any client.

fotoetienne reviewed (commented):
Great work! Made a couple revisions/suggestions

aprilrd committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Stephen Spalding <fotoetienne@users.noreply.github.com>

fotoetienne committed:
Formally specify  Non-Null behavior

This edit adds a distinct Behavior section that specifies how a client-side Non-Null operator functions

This is separated from the Syntax proposal to avoid confusion.

fotoetienne committed:
Added discussion on partial results

fotoetienne committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <fabiobean2@gmail.com>

fotoetienne committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <fabiobean2@gmail.com>

fotoetienne committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Alex Reilly <fabiobean2@gmail.com>

fotoetienne committed:
typos

twof committed:
Merge pull request #5 from fotoetienne/QueryLevelNullability

Formally specify  Non-Null behavior

mjmahone commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,348 @@
+
+While this solution simplifies some client-side logic, it does not meaningfully improve the developer experience for clients.
+
+* The cache implementations of GraphQL client libraries also need to understand the custom directive to behave correctly. Currently, when a client library caches a null field based on an operation without a directive, it will return the null field for another operation with this directive.
```
This will be true even with custom syntax. For composability of fragments, it would be worrisome if this new syntax made it impossible to have these two fragments in the same application:
```
fragment UserOkWithoutName on User {
  name
}
fragment UserMustHaveName on User {
  name!
}
```
With consistency, I could end up re-rendering a component built with `UserMustHaveName` based on a response fetched using only `UserOkWithoutName`.

And in fact it gets worse:
```
fragment BestFriendWithoutName on User {
  best_friend {
    id
  }
}
fragment BestFriendWithName on User {
  best_friend {
    id
    name!
  }
}
```
If my best friend, when I fetched `BestFriendWithName`, is `User:1`, and then suddenly, after fetching `BestFriendWithoutName` is `User:2`, I no longer am capable of rendering the component relying on `BestFriendWithName` (because `name` is *unset* on `User:2`). If the client library doesn't know what to do in that case, you can end up with consistency updates causing crashes.

aprilrd replied:
As you said, the client library will need to understand this new syntax/directive. I am not familiar with GraphQL client implementations, so I am not sure how difficult my imagined behavior will be to implement. Doesn't this behavior solve the issue you outlined?

> If a GraphQL client finds that a component relies on `BestFriendWithName `, but the cached data doesn't have `name`, it should not return any of the cached data. To the client, it is like a state before fetching started. Depending on `fetch-policy` (to use Apollo's terminology), the client can start a new request to fetch the missing field data.

mjmahone replied:
> Doesn't this behavior solve the issue you outlined?

That's how *Apollo* solves it, but that strategy isn't true universally. For instance that's not how Relay solves it. The client I work on *can't* solve it this way, as our consistency subscriptions are at the response level, for queries that include hundreds of fragments (whether that's a good choice or not is irrelevant: it's how things work for the largest GraphQL client at Facebook): if missing `name` meant the entire consistency update should not be served, whenever a product chose to mark their field with `!`, they could cause entire unrelated surfaces to lose consistency.

Another valid strategy would also be to return the data, but with `best_friend` being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for `@required(action: NONE)`). Or requiring the product to null check every piece of data, always, regardless of its nullability annotation.

aprilrd replied:
Do you mean that Relay requires field selections on the same id and type to behave the same globally? If so, I can see how my suggested behavior won't work. If not:

> Another valid strategy would also be to return the data, but with best_friend being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for @required(action: NONE))

This is what I imagined. The client would nullify up until the most immediate nullable parent.

lizjakubowski replied:
@mjmahone thank you for the feedback, this is great for us to think about! ❤️  I think I see the problem you've laid out here, but let me know if I've misunderstood anything!

when a not-required field is fetched and returned as `null`, we could end up wiping away values in the cache that are required by other features.

for features using codegen types, the Swift/Kotlin/etc. properties corresponding to those "required" fields will be non-Optional. the value will be replaced by `null` in the cache -- which would result in a crash anytime those non-Optional properties are accessed.

is it feasible for the client to distinguish in the response between these two cases (e.g. perhaps via the `errors` array)?
* fields that are `null` due to an error, and
* fields that are `null` because they are also `null` in the database

in the former case, the client library could choose to not clear the existing value in the cache. from some experimentation, this doesn't seem to be how e.g. the Apollo iOS cache works with regular nullable fields -- if a `null` value is returned, it seems to replace any value that was previously in the cache (I wasn't able to verify that this was also the case when an error occurs on that field).

is there any other way for the client library to know what to do in this case?

cometkim commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,348 @@
+Expressing nullability in the query, as opposed to the schema, offers the client more flexibility and control over
+whether or not an error is thrown.
+
+There are cases where a field is `nullable`, but a feature that fetches the field will not function if it is `null`.
+For example if you are trying to render an information page for a movie, you won't be able to do that if the name field
+of the movie is missing. In that case it would be preferable to fail as early as possible.
```
Here's one of the cases where this would be needed.

[GatsbyJS](https://www.gatsbyjs.com/) projects use GraphQL to handle build-time data dependencies, and provide a file system route API that can automatically build pages from specific types.

Creating a file called `src/pages/{BlogPost.slug}.jsx` creates pages from all BlogPost nodes and injects page context. `{ id: string, slug: string }`

user can use the page context as an argument to fetch additional data.

```graphql
# $id is automatically injected by Gatsby
query BlogPostQuery($id: String!) {
  # The type of blogPost is nullable, but the result is guaranteed to non-null by Gatsby
  blogPost(id: { eq: $id }) {
    title
    tags
  }
}
``` 

This can be done with additional static analysis in plugin layer. But the proposal can be a simple solution.

Even if It's not clear in production cases that can customize schema, but it seems a huge advantage for ecosystem players using SDL and code-generation.

aprilrd replied:
Thank you for sharing another use case!

twof committed:
Update QueryLevelNullability.md

twof committed:
Update QueryLevelNullability.md

twof committed:
Merge pull request #6 from twof/3rd-party-apis

Addendum about 3rd party APIs

josephsavona commented:
One general thought is that clients generally use GraphQL to fulfill at least two purposes in an application. 
* To describe the *local* data dependencies of an individual unit of UI in order read that data from the local cache (and subscribe for updates).
* To describe the *aggregated* data dependencies of an entire user interface in order to fetch those dependencies from the server.


An example of the distinction is pagination: 
* The user interface wants to read (from the local cache) all the items that have been fetched so far.
* The query sent to the server should always be for a page at a time: applications generally don't want to keep refetching from the beginning of the list (1-10, then 1-20, etc) but rather fetch pages (1-10, then 11-20, etc).


The same distinction is relevant when considering the "required-ness" of a field: Individual components want to describe their data requirements - if a given component must have `user.name`, then it should be easy to express that dependency. However, when data dependencies of multiple components are aggregated that should occur in a way that allows fetching data to fulfill as much of the user interface as possible. 

**This proposal makes it difficult to use the same GraphQL to describe an individual component's data dependencies _and_ to aggregate the dependencies for an entire user interface**. Instead, the likely outcome of this proposal is that "sophisticated" clients (ie those that use AOT query compilation and also that use a client-side cache as a source of truth) would use `!` as both a user-facing feature and as a compiler target. In other words:
* When reading data for a component from the cache, `!` would allow a component to express that a field is required and have the cache return "null" for an entire fragment (or part of one) if a required field was missing _in the cache_.
* When sending queries to the server, **the client would generally remove `!`** unless it could prove that a field was *always* used with `!`, or more generally that the inclusion of `!` on a field would not break fragment composition.

Considering this last point, it is likely that under this proposal many instances of `!` fields would likely end up pruned by client compilers, losing information about potential optimizations that could be applied. A canonical failure case is below (adapted from @mjmahone's comment above):

```
query UserSummary {
  me {
    ...Header
    ...Timeline
   }
 }
 
fragment Header on User {
  name!
  profile_picture
  ...SomeExpensiveFragment
}

fragment Timeline on User {
  name
  posts(first: 10)! { 
    title!  # exclude posts that don't have a title
  }
}

fragment SomeExpensiveFragment on User { ... }
```

Imagine that the user interface is capable of rendering Header and Timeline independently: even if data for one is missing the application should render the section whose data is available. If `me.name` is null at runtime, then we would ideally do the following:
* Continue to resolve all the selections of `Timeline`: its data dependencies can still be fulfilled w a null `name`
* Avoid (or cancel) loading `SomeExpensiveFragment`, since we know that `Header` cannot be fulfilled.

In practice, a client compiler would have to transform the query as follows:

```
...snip...

fragment Header on User {
  name # <--- remove `!` here because keeping it could cause `Timeline` to fail
  ...snip...
}
fragment Timeline on User {
  ...snip...
  posts(first: 10) {  # <-- remove `!` here because keeping it could `Header` to fail
    ...snip...
  }
}
...snip...
```

In other words: _client compilers would have to remove both instances of `!` in the version of the query that they send to the server_. At the same time, there is no way for the client to describe its real intent, which is that if `Header.name` cannot be fulfilled, the remaining selections of `Header` can be skipped, and similarly if `Timeline.posts` is null, that the other selections of `Timeline` can be skipped. I'd encourage thinking about ways to retain the ability to declare "required-ness" of a field but potentially change the semantics to accommodate these cases.

twof committed:
Cleanup after behavior was clarified

twof committed:
Remove a few instances of query when they should be operation

twof committed:
revert non-null

twof committed:
Merge pull request #7 from twof/behavior-cleanup

Cleanup after behavior was clarified

xuewei8910 commented on rfcs/QueryLevelNullability.md:
```diff
@@ -0,0 +1,342 @@
+- [Mark Larah](https://github.com/magicmark) - Yelp Web
+- [Sanae Rosen](<social or github link here>) - Yelp Android
+- [Stephen Spalding](https://github.com/fotoetienne) - Netflix GraphQL Server Infrastructure
+- [Wei Xue](<social or github link here>) - Yelp iOS
```
```suggestion
- [Wei Xue](https://github.com/xuewei8910) - Yelp iOS
```

twof committed:
Update rfcs/QueryLevelNullability.md

Co-authored-by: Wei Xue <xuew8910@gmail.com>

twof commented:
Thanks everyone for your comments! Earlier today the champions of this proposal attended a GraphQL Working Group meeting and officially [made it to the first RFC stage](https://github.com/graphql/graphql-spec/issues/867#issuecomment-840807186). You can read [the notes from the meeting here](https://docs.google.com/document/d/1fNjlQdZ6Y5So6d3hQmGC_O9g-9CNp_qgDhPRRPNd2uw/edit#heading=h.pewbya360jh9), and I'll update again when the recording of the presentation is online. 

We'll be moving this over to a PR in the main repo, and I'll try so summarize the existing comments as best I can. We definitely still want to address them.
